crates/bevy_reflect/src/path/access.rs

```diff
@@ -0,0 +1,225 @@
+use std::{borrow::Cow, fmt};
+
+use super::{AccessError, ReflectPathError};
+use crate::{Reflect, ReflectMut, ReflectRef, VariantType};
+use thiserror::Error;
+
+type InnerResult<T> = Result<Option<T>, Error<'static>>;
+
+#[derive(Debug, PartialEq, Eq, Error)]
+pub(super) enum Error<'a> {
+    #[error(
+        "the current {ty} doesn't have the {} {}",
+        access.kind(),
+        access.display_value(),
+    )]
+    Access { ty: TypeShape, access: Access<'a> },
+
+    #[error("invalid type shape: expected {expected} but found a reflect {actual}")]
```

Comment:
Replying to [this](https://github.com/bevyengine/bevy/pull/8887/files#r1252677220) comment:
> yeah, but `Type` here is not really a "type" but a "type shape", The error message would read as follow: "expected struct but found tuple". I don't think there is such a thing as a "struct-like" it's either a struct or not a struct.

Well I think it would be more useful for `TypeShape::Tuple` and `TypeShape::List`, since those also work for tuple structs and arrays, respectively. It's not required we use `-like` haha, I just think it would be clearer to users that we're not expecting them to pass only `Struct`, `Tuple`, and `List` types.